nltk/chunk/regexp.py

```diff
@@ -1217,11 +1219,12 @@ def _read_grammar(self, grammar, root_label, trace):
         """
         rules = []
         lhs = None
+        pattern = regex.compile("(?P<nonterminal>(\\.|[^:])*)(:(?P<rule>.*))")
         for line in grammar.split("\n"):
             line = line.strip()
 
             # New stage begins if there's an unescaped ':'
-            m = re.match("(?P<nonterminal>(\\.|[^:])*)(:(?P<rule>.*))", line)
+            m = pattern.match(line)
             if m:
                 # Record the stage that we just completed.
                 self._add_stage(rules, lhs, root_label, trace)
```

Comment:
`regex` to avoid exponential backtracking. Compile outside of the for loop for efficiency.